nucleus/flashlight/framework/src/main/java/org/glassfish/flashlight/impl/client/ProbeProviderClassFileTransformer.java

```diff
@@ -240,37 +190,35 @@ final synchronized void untransform() {
                 return null;
             }
 
-            if (enabled) {
+            if (hasEnabledInvokers()) {
                 // we still need to write out the class file in debug mode if it is
                 // disabled.
-                cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES + ClassWriter.COMPUTE_MAXS);
+                ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES + ClassWriter.COMPUTE_MAXS);
                 ClassReader cr = new ClassReader(classfileBuffer);
                 cr.accept(new ProbeProviderClassVisitor(cw), null, 0);
-                ret = cw.toByteArray();
+                byte[] instrumentedClassBytes = cw.toByteArray();
                 Log.fine("transformed", providerClassName);
                 if (_debug) {
                   ProbeProviderClassFileTransformer.writeFile(className.substring(className.lastIndexOf('/') + 1)+"supplied_"+count, classfileBuffer);
-                  ProbeProviderClassFileTransformer.writeFile(className.substring(className.lastIndexOf('/') + 1)+"transformed_"+count, ret);
+                  ProbeProviderClassFileTransformer.writeFile(className.substring(className.lastIndexOf('/') + 1)+"transformed_"+count, instrumentedClassBytes);
                   count++;
                 }
+                return instrumentedClassBytes;
             }
-            else {
-                if (_debug) {
-                  ProbeProviderClassFileTransformer.writeFile(className.substring(className.lastIndexOf('/') + 1)+"supplied_"+count, classfileBuffer);
-                  count++;
-                }
-                ret = null;
-                Log.fine("untransformed", providerClass.getName());
+            if (_debug) {
+              ProbeProviderClassFileTransformer.writeFile(className.substring(className.lastIndexOf('/') + 1)+"supplied_"+count, classfileBuffer);
+              count++;
             }
+            Log.fine("untransformed", providerClass.getName());
+            return null;
         }
         catch (Exception ex) {
             logger.log(Level.WARNING, REGISTRATION_ERROR, ex);
+            return null;
         }
-
-        return ret;
     }
 
-    private synchronized void addTransformer() {
+    private void addTransformer() {
```

Comment:
Since all updates are run by the same thread no synchronisation is needed any more.